lib/options.c

```diff
@@ -40,371 +40,249 @@
+#ifdef HAVE_TCTI_DEV
+#include "tpm2_tools_tcti_device.h"
 #endif
-    default:
-        fprintf (stderr, "invalid TCTI, see --help\n");
-        return 2;
+
+/* Environment variables usable as alternatives to command line options */
+#define TPM2TOOLS_ENV_TCTI_NAME      "TPM2TOOLS_TCTI_NAME"
+
+struct tpm2_options {
+    tpm2_option_handler on_opt;
+    tpm2_arg_handler on_arg;
+    char *short_opts;
+    size_t len;
+    struct option long_opts[];
+};
+
+tpm2_options *tpm2_options_new(const char *short_opts, size_t len, struct option *long_opts, tpm2_option_handler on_opt, tpm2_arg_handler on_arg) {
+
+    tpm2_options *opts = calloc(1, sizeof(*opts) + (sizeof(*long_opts) * len));
+    if (!opts) {
+        LOG_ERR("oom");
+        return NULL;
     }
-    return 0;
-}
-/*
- * Populate the provided common_opts_t structure with data provided through
- * the environment.
- */
-void
-get_common_opts_from_env (common_opts_t *common_opts)
-{
-    char *env_str, *end_ptr;
-
-    if (common_opts == NULL)
-        return;
-    env_str = getenv (TPM2TOOLS_ENV_TCTI_NAME);
-    if (env_str != NULL)
-        common_opts->tcti_type = tcti_type_from_name (env_str);
-    env_str = getenv (TPM2TOOLS_ENV_DEVICE_FILE);
-    if (env_str != NULL)
-        common_opts->device_file = env_str;
-    env_str = getenv (TPM2TOOLS_ENV_SOCKET_ADDRESS);
-    if (env_str != NULL)
-        common_opts->socket_address = env_str;
-    env_str = getenv (TPM2TOOLS_ENV_SOCKET_PORT);
-    if (env_str != NULL)
-        common_opts->socket_port = strtol (env_str, &end_ptr, 10);
+
+    opts->short_opts = strdup(short_opts);
+    if (!opts) {
+        LOG_ERR("oom");
+        free(opts);
+        return NULL;
+    }
+
+    opts->on_opt = on_opt;
+    opts->on_arg = on_arg;
+    opts->len = len;
+    memcpy(opts->long_opts, long_opts, len * sizeof(*long_opts));
+
+    return opts;
 }
-/*
- * Append a string to the parameter argv array. We realloc this array adding
- * a new char* to point to the appended entry. The argc parameter is updated
- * to account for the new element in the array. We return the address of the
- * newly reallocated array. If the realloc fails we return NULL.
- */
-char**
-append_arg_to_vector (int*  argc,
-                      char* argv[],
-                      char* arg_string)
-{
-    char **new_argv;
-
-    ++(*argc);
-    new_argv = realloc (argv, sizeof (char*) * (*argc));
-    if (new_argv != NULL) {
-        new_argv[*argc - 1] = arg_string;
-    } else {
-        LOG_ERR("Failed to realloc new_argv to append string %s: %s",
-                arg_string,
-                strerror (errno));
+
+#define OPTIONS_SIZE(x) (sizeof(*x) + (sizeof(x->long_opts) * x->len))
+tpm2_options *tpm2_options_cat(tpm2_options *a, tpm2_options *b) {
+
+    size_t long_opts_len = a->len + b->len;
+    /* +1 for a terminating NULL at the end of options array for getopt_long */
+    tpm2_options *tmp = calloc(1, sizeof(*a) + 1 + (long_opts_len * sizeof(a->long_opts[0])));
```

Comment:
I guess you defined the OPTIONS_SIZE() to use here but forgot?

Reply:
But I guess you instead want:

#define OPTIONS_SIZE(opt, len) (sizeof(*opt) + (sizeof(opt->long_opts[0]) * len))

tpm2_options *tmp = calloc(1, OPTIONS_SIZE(a, long_opts_len)  + 1);